File: April 2020/28 - First Unique Number/Solution.py
from collections import defaultdict, deque 

class FirstUnique:

    def __init__(self, nums: List[int]):
        self._mp = defaultdict()
        self._q = []
        
        for num in nums:
            self.add(num)

    def showFirstUnique(self) -> int:
        while self._q and self._mp[_q[0]] > 1:
            self._q.pop(0)
        
        if self._q:
            return self._q[0]
        return -1

    def add(self, value: int) -> None:
        self._q.append(value)
        self._mp[value] += 1

File: April 2020/28 - First Unique Number/Solution.cpp
class FirstUnique {
    
    unordered_map<int, int> _mp;
    queue<int> _q;
    
public:
    FirstUnique(vector<int>& nums) {
        for(int num : nums)
            add(num);
    }
    
    int showFirstUnique() {
        while(!_q.empty() && _mp[_q.front()] > 1)
            _q.pop();
        
        if(_q.empty())
            return -1;
        return _q.front();
    }
    
    void add(int value) {
        _q.push(value);
        _mp[value]++;
    }
};
